Tests fail if run from under a target directory

## Expected behavior
As part of rewrite-clj canary testing, I extract a copy of refactor-nrepl source code somewhere beneath rewrite-clj's `target` directory.

When I run 
```
make test
```
I should see all tests pass.

## Actual behavior
I see the following test failures:
```
...

FAIL in (source-files-with-clj-like-extension-test) (core_test.clj:45)
expected: (pos? (count (filter (fn [f] (-> f .getPath (.endsWith extension))) result)))
  actual: (not (pos? 0))

lein test :only refactor-nrepl.core-test/source-files-with-clj-like-extension-test

FAIL in (source-files-with-clj-like-extension-test) (core_test.clj:45)
expected: (pos? (count (filter (fn [f] (-> f .getPath (.endsWith extension))) result)))
  actual: (not (pos? 0))

lein test :only refactor-nrepl.core-test/source-files-with-clj-like-extension-test

FAIL in (source-files-with-clj-like-extension-test) (core_test.clj:45)
expected: (pos? (count (filter (fn [f] (-> f .getPath (.endsWith extension))) result)))
  actual: (not (pos? 0))

...

FAIL in (sorts-by-frequencies) (form-init11474675976321826582.clj:32)
expected: (= (first utils) (quote refactor-nrepl.util))
  actual: (not (= nil refactor-nrepl.util))

lein test :only refactor-nrepl.ns.namespace-aliases-test/libspecs-are-cached

FAIL in (libspecs-are-cached) (form-init11474675976321826582.clj:42)
expected: (thrown-with-msg? Exception #"Expected!" (sut/namespace-aliases false))
  actual: nil

...

Ran 130 tests containing 29599 assertions.
5 failures, 0 errors.
Tests failed.
Error encountered performing task 'test' with profile(s): 'base,system,provided,dev,1.10,config'
Tests failed.
make: *** [test] Error 1
```

If I instead don't have `target` in my path all tests pass.

## Steps to reproduce the problem
```
mkdir -p foobar/target
cd foobar/target
git clone https://github.com/clojure-emacs/refactor-nrepl.git
cd refactor-nrepl
git reset v3.0.0 --hard
make test
```
You will see tests fail, but if you then:
```
cd ../..
mv target ok
cd ok/refactor-nrepl
make test
```
You will see tests pass.

## Environment & Version information

### clj-refactor.el and refactor-nrepl version information
Running tests from v3.0.0 tag clone of this repository.

### CIDER version information
n/a

### Leiningen or Boot version
```
❯ lein --version
Leiningen 2.9.7 on Java 11.0.13 OpenJDK 64-Bit Server VM
```
### Emacs version
n/a

### Operating system
macOS 10.15.7

## Diagnosis
This was working for refactor-nrepl 2.5.1.

I think the issue is likely caused by new classpath exclusion techniques:
https://github.com/clojure-emacs/refactor-nrepl/blob/a3f9de809b1246101786ec6160df3b8c66585f74/src/refactor_nrepl/core.clj#L62-L72

which excludes any path that contains `target` or `resources` (I'm guessing that `.gitlibs` is less important).

I raised this issue in case it might be affecting your broader user base.
If it is only affecting my particular use case, I can find a work-around for rewrite-clj canary testing.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,7 +5,8 @@
 ### Changes
 
 * [#344](https://github.com/clojure-emacs/refactor-nrepl/issues/344): make clean-ns's style closer to the [how to ns](https://stuartsierra.com/2016/08/27/how-to-ns) style.
-* [#333](https://github.com/clojure-emacs/refactor-nrepl/issues/333): skip scanning irrelevant directories in more places (as it was already done for various other functionalities). 
+* [#333](https://github.com/clojure-emacs/refactor-nrepl/issues/333): skip scanning irrelevant directories in more places (as it was already done for various other functionalities; this limits ananlysis/refactoring to your source/test paths, skipping other artifacts). 
+* [#346](https://github.com/clojure-emacs/refactor-nrepl/issues/346): refine the heuristic for ignoring irrelevant dirs (see the above bullet point).
 * Introduce `print-right-margin`/`print-miser-width` configuration options, used during `pprint`ing of ns forms.
   * The default is one that is consistent with refactor-nrepl's traditional behavior.
   * You can set both to `nil` for disabling line wrapping. 

--- a/src/refactor_nrepl/core.clj
+++ b/src/refactor_nrepl/core.clj
@@ -65,18 +65,28 @@
   Should generally always be true, except in a few refactor-nrepl's unit tests."
   true)
 
+(def ^:private resources-dir-pattern
+  (re-pattern (str "resources" File/separator "?$")))
+
+(def ^:private target-dir-pattern
+  (re-pattern (str "target" File/separator "?$")))
+
+(defn irrelevant-dir?
+  "Does `f` look like a directory without files that should be analyzed/refactored?"
+  [^File f]
+  (let [s (-> f .toString)]
+    (boolean (or (if *skip-resources?*
+                   (re-find resources-dir-pattern s)
+                   false)
+                 (re-find target-dir-pattern s)
+                 (-> s (.contains ".gitlibs"))))))
+
 (defn source-dirs-on-classpath
   "Like `#'dirs-on-classpath`, but restricted to dirs that look like
   (interesting) source/test dirs."
   []
   (->> (dirs-on-classpath)
-       (remove (fn [^File f]
-                 (let [s (-> f .toString)]
-                   (or (if  *skip-resources?*
-                         (-> s (.contains "resources"))
-                         false)
-                       (-> s (.contains "target"))
-                       (-> s (.contains ".gitlibs"))))))
+       (remove irrelevant-dir?)
        (remove util/dir-outside-root-dir?)))
 
 (defn project-root

--- a/test/refactor_nrepl/core_test.clj
+++ b/test/refactor_nrepl/core_test.clj
@@ -45,3 +45,27 @@
       (is (pos? (count (filter (fn [^File f]
                                  (-> f .getPath (.endsWith extension)))
                                result)))))))
+
+(deftest irrelevant-dir?
+  (are [input expected] (= expected
+                           (sut/irrelevant-dir? (File. input)))
+    "resources"         true
+    "a/resources"       true
+    "a/resources/"      true
+    "a/resources/b"     false
+
+    "dev-resources"     true
+    "a/dev-resources"   true
+    "a/dev-resources/"  true
+    "a/dev-resources/b" false
+
+    "target"            true
+    "a/target"          true
+    "a/target/"         true
+    "a/target/b"        false
+
+    ;; if the string ".gitlibs" is contained, it's always irrelevant:
+    ".gitlibs"          true
+    "a/.gitlibs"        true
+    "a/.gitlibs/"       true
+    "a/.gitlibs/b"      true))

--- a/test/refactor_nrepl/ns/resolve_missing_caching_test.clj
+++ b/test/refactor_nrepl/ns/resolve_missing_caching_test.clj
@@ -1,10 +1,10 @@
 (ns refactor-nrepl.ns.resolve-missing-caching-test
-  (:require [clojure
-             [edn :as edn]]
-            [nrepl.core :as nrepl]
-            [refactor-nrepl.ns.resolve-missing-test :refer [session-fixture]]
-            [clojure.test :refer [deftest testing is use-fixtures]]
-            [clojure.string :as str]))
+  (:require
+   [clojure.edn :as edn]
+   [clojure.string :as str]
+   [clojure.test :refer [deftest testing is use-fixtures]]
+   [nrepl.core :as nrepl]
+   [refactor-nrepl.ns.resolve-missing-test :refer [session-fixture]]))
 
 (use-fixtures :each session-fixture)
 

